library/Icingadb/Web/Controller.php

```diff
@@ -550,6 +550,86 @@ public function export(Query ...$queries)
             return true;
         }
 
+        if ($this->format === 'json' || $this->format === 'csv') {
+            $isJsonFormat = $this->format === 'json';
+            $response = $this->getResponse();
+
+            $fileName = $this->getRequest()->getControllerName();
+            if ($isJsonFormat) {
+                $response
+                    ->setHeader('Content-Type', 'application/json')
+                    ->setHeader('Cache-Control', 'no-store')
+                    ->setHeader(
+                        'Content-Disposition',
+                        'inline; filename=' . $fileName . '.json'
+                    )
+                    ->sendResponse();
+            } else {
+                $response
+                    ->setHeader('Content-Type', 'text/csv')
+                    ->setHeader('Cache-Control', 'no-store')
+                    ->setHeader(
+                        'Content-Disposition',
+                        'attachment; filename=' . $fileName . '.csv'
+                    )
+                    ->sendResponse();
+            }
+            // Disable output buffering to stream large amounts of data without a memory exhausted error.
+            ob_end_clean();
+            set_time_limit(0);
+
+            $query = $queries[0];
+            $query->limit(Url::fromRequest()->getParam('limit'));
+            $query->resetOffset();
+            $this->applyRestrictions($query);
+
+            $keyName = is_array($query->getModel()->getKeyName())
+                ? $query->getModel()->getKeyName()
+                : (array) $query->getModel()->getKeyName();
+            $col = array_merge($keyName, $query->getModel()->getColumns());
```

Comment:
Simplify this to `$col = array_merge((array) $query->getModel()->getKeyName(), ...` please.

And don't export only the base model's columns. There may be more models (tables) joined in and these can and should be exported as well. Identify them with `$query->getWith()`. Joined table columns should be prefixed with the paths that are returned by it.